src/librustc_typeck/collect.rs

```diff
@@ -2211,7 +2212,12 @@ fn predicates_from_bound<'tcx>(
     match *bound {
         hir::GenericBound::Trait(ref tr, hir::TraitBoundModifier::None) => {
             let mut bounds = Bounds::default();
-            let (pred, _) = astconv.instantiate_poly_trait_ref(tr, param_ty, &mut bounds);
+            let (pred, _) = astconv.instantiate_poly_trait_ref(
+                tr,
+                param_ty,
+                &mut bounds,
+                false,
```

Comment:
Would be good to leave a comment for this one since it's the only case where `include_atb == false`.

Reply:
Yep. Will do so once I consult with Niko, per the above comment. :-)

Reply:
OK, so, it was not very obvious to me why passing `false` here makes sense -- but I believe it is true, given where this function is called from. Perhaps we can adust the name. Essentially, it only makes sense in those cases where we are assembling the bounds only for the purpose of resolving `T::Bar` to a specific trait. 

We definitely need to adjust the names and apply some comments -- but not just here, also to the callers. The key point is that we are in a context where we are trying to resolve `T::Bar`, and we only want *just as many* predicates as we need to do that (to avoid artificial cycles).

I'm not sure if adding this `false` parameter is the best approach or not, but I don't immediately see a better one.

Reply:
(I just spent a while tracing back through the code, checking on things)